checks/permissions.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2021 Security Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checks
+
+import (
+	"errors"
+	"fmt"
+
+	"gopkg.in/yaml.v2"
+
+	"github.com/ossf/scorecard/checker"
+)
+
+const CheckPermissions = "Token-Permissions"
+
+// ErrInvalidGitHubWorkflowFile : Invalid GitHub workflow file.
+var ErrInvalidGitHubWorkflowFile = errors.New("invalid GitHub workflow file")
+
+//nolint:gochecknoinits
+func init() {
+	registerCheck(CheckPermissions, leastPrivilegedTokens)
+}
+
+func leastPrivilegedTokens(c *checker.CheckRequest) checker.CheckResult {
+	return CheckFilesContent(CheckPermissions, ".github/workflows/*", true, c, validateGitHubActionTokenPermissions)
+}
+
+func validatePermission(key string, value interface{}, path string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	switch val := value.(type) {
+	case string:
+		if val == "write" {
+			logf("!! token-permissions/github-token - %v permission set to '%v' in %v", key, val, path)
+			return false, nil
+		}
+	default:
+		return false, ErrInvalidGitHubWorkflowFile
+	}
+	return true, nil
+}
+
+func validateMapPermissions(values map[interface{}]interface{}, path string,
+	logf func(s string, f ...interface{})) (bool, error) {
+	permissionRead := true
+	var r bool
+	var err error
+
+	// Iterate over the permission, verify keys and values are strings.
+	for k, v := range values {
+		switch key := k.(type) {
```

Comment:
probably simpler as just:

```go
key, ok := k.(string)
if !ok {
  return false, ErrInvalidGitHubWorkflowFile
}

if r, err = validatePermission(key, v, path, logf); err != nil {
    return false, err
}
...
```

Reply:
sorry, did not see before merging. I created #539 to address it.